### PR TITLE
Generate an instrumented vDVS plugin binary

### LINKED ISSUE #1589: Generate an instrumented vDVS plugin binary
As per https://github.com/golang/go/issues/16537#issuecomment-303218184 ; `go build` is not supporting to generate instrumented binary to capture coverage for plugin code. The approach mentioned at [external testing](https://www.elastic.co/blog/code-coverage-for-your-golang-system-tests) stands out as workaround until go tooling supports by default. 

To achieve code coverage for vDVS plugin, we have to follow the workaround mentioned at above external testing link to come up with instrumented vDVS test-coverage binary.

### PR DESCRIPTION
**Pull request by shuklanirdesh82**
Fixes #1589 

Through this PR, we are able to generate instrumented vDVS binary and will be placed at `build` folder. 

Testing done: Tested locally

```
make clean build-all
go tool vet .
gofmt -s -l -w .
go build --ldflags '-extldflags "-static"' -o ../build/docker-volume-vsphere github.com/vmware/docker-volume-vsphere/client_plugin/vmdk_plugin
go test -c -o ../build/vmdkops.test github.com/vmware/docker-volume-vsphere/client_plugin/drivers/vmdk/vmdkops -cover
go test -c -o ../build/docker-volume-vsphere.test github.com/vmware/docker-volume-vsphere/client_plugin/vmdk_plugin -cover -tags sanity
go test -coverprofile=/tmp/cover.out -coverpkg=github.com/vmware/docker-volume-vsphere/client_plugin/... -c -o ../build/vdvs-instrumented github.com/vmware/docker-volume-vsphere/client_plugin/vmdk_plugin -tags r -covermode count
warning: no packages being tested depend on github.com/vmware/docker-volume-vsphere/client_plugin/drivers/shared
warning: no packages being tested depend on github.com/vmware/docker-volume-vsphere/client_plugin/shared_plugin
make  --directory=../esx_service build
```
Following steps were performed to test out generated binary
1. make clean build-all
2. copying `build/vdvs-instrumented` to docker host
3. run instrumented binary instead of installing plugin
`./vdvs-instrumented -test.coverprofile=e2e.cov &`
4. Run some basic test `make test-e2e -test.run "*Basic*"` (locally changed the tag to pick up basic test only)
5. hist `ctrl+c` to end the program (on the docker host where step#2 is performed)
6. look for `e2e.cov` is generated into the current directory

later `e2e.cov` would be passed to `go tool cover` to analyze further.

Note: `e2e.cov` is renamed as `e2e.txt` to upload here. [e2e.txt](https://github.com/vmware/docker-volume-vsphere/files/1174696/e2e.txt)

### CODE REVIEW

File: client_plugin/vmdk_plugin/sanity_test.go [hunk 1]
```diff
@@ -16,6 +16,8 @@
 // VMDK Docker driver sanity tests.
```
Comment (lipingxue) [inline client_plugin/vmdk_plugin/sanity_test.go]: Could you add some comments about what is the sanity test is doing?

File: client_plugin/Makefile [hunk 2]
```diff
@@ -153,12 +160,15 @@ $(BIN)/$(VMDKOPS_TEST_MODULE).test: $(VMDKOPS_MODULE_SRC) $(TEST_SRC) $(VMDKOPS_
...
 
 $(BIN)/$(PLUGNAME).test: $(COMMON_SRC) $(VMDK_PLUGIN_SRC) $(TEST_SRC) ./vmdk_plugin/*_test.go
-	$(GO) test -c -o $@ $(PLUGIN)/vmdk_plugin -cover
+	$(GO) test -c -o $@ $(PLUGIN)/vmdk_plugin -cover -tags sanity
 
 $(SHARED_PLUGIN_BIN): $(COMMON_SRC) $(SHARED_PLUGIN_SRC)
 	@-mkdir -p $(BIN) && chmod a+w $(BIN)
 	$(GO) build --ldflags '-extldflags "-static"' -o $(SHARED_PLUGIN_BIN) $(PLUGIN)/shared_plugin
 
+$(BIN)/$(INSTRUMENTED_PLUGIN_BIN): $(COMMON_SRC) $(VMDKOPS_MODULE_SRC) $(E2E_TEST_SRC) ./vmdk_plugin/*_test.go
```
Comment (lipingxue) [inline client_plugin/Makefile]: Please define "VMDK_PLUGIN_TEST_SRC"=./vmdk_plugin/*_test.go" here and use $(VMDK_PLUGIN_TEST_SRC)".

File: client_plugin/Makefile [hunk 3]
```diff
@@ -72,7 +73,7 @@ PLUGIN_BIN = $(BIN)/$(PLUGNAME)
 SHARED_PLUGIN_BIN = $(BIN)/$(SHARED_PLUGNAME)
 
 # all binaries for VMs - plugin and tests
-VM_BINS = $(PLUGIN_BIN) $(BIN)/$(VMDKOPS_TEST_MODULE).test $(BIN)/$(PLUGNAME).test
+VM_BINS = $(PLUGIN_BIN) $(BIN)/$(VMDKOPS_TEST_MODULE).test $(BIN)/$(PLUGNAME).test $(BIN)/$(INSTRUMENTED_PLUGIN_BIN)
```
Comment (msterin) [inline client_plugin/Makefile]: pls add a comment about "instrumented for <what is it instrumented for?"

File: client_plugin/Makefile [hunk 4]
```diff
@@ -105,6 +106,12 @@ SHARED_PLUGIN_SRC = shared_plugin/main.go drivers/shared/shared_driver.go
 
 TEST_SRC = ../tests/utils/inputparams/testparams.go
 
+E2E_TEST_SRC = ../tests/utils/admincli/*.go ../tests/utils/dockercli/*.go \
```
Comment (msterin) [inline client_plugin/Makefile]: should it be ../tests/{utils,constants}/*/*.go ../test/e2e/*.go ?

File: client_plugin/vmdk_plugin/sanity_test.go (hunk 1 above)
Comment (shuklanirdesh82) [inline client_plugin/vmdk_plugin/sanity_test.go]: this test was left out with the tags .. so adding a tag for the test to control during the build time.

File: client_plugin/Makefile (hunk 2 above)
Comment (shuklanirdesh82) [inline client_plugin/Makefile]: done!

File: client_plugin/Makefile (hunk 3 above)
Comment (shuklanirdesh82) [inline client_plugin/Makefile]: done!

File: client_plugin/Makefile (hunk 4 above)
Comment (shuklanirdesh82) [inline client_plugin/Makefile]: Well it is not needed, I have removed. 

Thanks for pointing out!

**Comment by pshahzeb** [comment]
@shuklanirdesh82 with this instrumented binary will we be able to use -check.v flag when run tests (as go test -check.v) so that verbose details of which test passed and failed inside a test suite are printed?

**Comment by shuklanirdesh82** [comment]
> https://github.com/vmware/docker-volume-vsphere/pull/1633#issuecomment-318788732

No @pshahzeb why do you think so?

**Comment by pshahzeb** [comment]
As per the offline discussion we had had over the issue of #1509, I was wondering if it would be possible to do so.

**Pull request merged**